clients/unshaded/src/main/java/tachyon/client/netty/NettyRemoteBlockReader.java

```diff
@@ -0,0 +1,132 @@
+import tachyon.network.ChannelType;
+import tachyon.network.NettyUtils;
+import tachyon.network.protocol.RPCBlockRequest;
+import tachyon.network.protocol.RPCBlockResponse;
+import tachyon.network.protocol.RPCMessage;
+import tachyon.network.protocol.RPCMessageDecoder;
+import tachyon.network.protocol.RPCMessageEncoder;
+import tachyon.network.protocol.RPCResponse;
+
+/**
+ * Read data from remote data server using Netty.
+ */
+public final class NettyRemoteBlockReader implements RemoteBlockReader {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  // Share both the encoder and decoder with all the client pipelines.
+  private static final RPCMessageEncoder ENCODER = new RPCMessageEncoder();
+  private static final RPCMessageDecoder DECODER = new RPCMessageDecoder();
+
+  private static final TachyonConf TACHYON_CONF = new TachyonConf();
+  private static final ChannelType CHANNEL_TYPE = TACHYON_CONF.getEnum(
+      Constants.USER_NETTY_CHANNEL, ChannelType.defaultType());
+  private static final Class<? extends SocketChannel> CLIENT_CHANNEL_CLASS = NettyUtils
+      .getClientChannelClass(CHANNEL_TYPE);
+  // Reuse EventLoopGroup for all clients.
+  // Use daemon threads so the JVM is allowed to shutdown even when daemon threads are alive.
+  private static final EventLoopGroup WORKER_GROUP = NettyUtils.createEventLoop(CHANNEL_TYPE,
+      TACHYON_CONF.getInt(Constants.USER_NETTY_WORKER_THREADS, 1), "netty-client-worker-%d", true);
+
+  private final Bootstrap mClientBootstrap;
+  private final ClientHandler mHandler;
+
+  // TODO: Creating a new remote block reader may be expensive, so consider a connection pool.
+  public NettyRemoteBlockReader() {
+    mHandler = new ClientHandler();
+    mClientBootstrap = createClientBootstrap();
+  }
+
+  @Override
+  public ByteBuffer readRemoteBlock(String host, int port, long blockId, long offset, long length)
+      throws IOException {
+    InetSocketAddress address = new InetSocketAddress(host, port);
+
+    try {
+      ChannelFuture f = mClientBootstrap.connect(address).sync();
+
+      LOG.info("Connected to remote machine " + address);
+      Channel channel = f.channel();
+      SingleResponseListener listener = new SingleResponseListener();
+      mHandler.addListener(listener);
+      channel.writeAndFlush(new RPCBlockRequest(blockId, offset, length));
+
+      RPCResponse response = listener.get(1, TimeUnit.SECONDS);
+      channel.close().sync();
+
+      if (response.getType() == RPCMessage.Type.RPC_BLOCK_RESPONSE) {
+        RPCBlockResponse blockResponse = (RPCBlockResponse) response;
+        LOG.info("Data " + blockId + " from remote machine " + address + " received");
+
+        if (blockResponse.getBlockId() < 0) {
+          LOG.info("Data " + blockResponse.getBlockId() + " is not in remote machine.");
+          return null;
+        }
+        return blockResponse.getPayloadDataBuffer().getReadOnlyByteBuffer();
+      }
```

Comment:
better to have an `else` case if  `response.getType() != RPCMessage.Type.RPC_BLOCK_RESPONSE`, even a single `LOG.error()`